gcc/rust/backend/rust-compile-item.h

```diff
@@ -138,10 +138,19 @@ class CompileItem : public HIRCompileBase
     // convert to the actual function type
     auto compiled_fn_type = TyTyCompile::compile (ctx->get_backend (), fnType);
 
+    unsigned int flags = 0;
+    bool is_main_fn = function.function_name.compare ("main") == 0;
+
+    // if its the main fn or pub visibility mark its as DECL_PUBLIC
+    // please see https://github.com/Rust-GCC/gccrs/pull/137
+    if (is_main_fn || function.has_visibility ())
+      flags |= Backend::function_is_visible;
+
+    // FIXME need name mangling
+    std::string asm_name = function.function_name;
```

Comment:
The `main` function doesn't need the name mangling. I mention it here in case you miss it.

Reply:
The rust main function needs symbol mangling. Only the main shim doesn't.

Reply:
Alright, I see.
In Rust, the `main` function will be mangling first, then connect to the main shim. And the main shim was generated as a convention automatically.
This is different from C/C++, the `main` function is always the default symbol linked to `_start` defined in the linker script.

Thanks for correcting me! @bjorn3

Reply:
Yeah, libstd does some extra work in addition to what the libc startup does like setting up a stack guard, adding a handler for SIGSEGV to give a nice error message on stack overflows, storing the program arguments in a static and running the main function within `std::panic::catch_unwind` to ensure that panicking doesn't immediately abort due to a missing unwind catcher. For reference the main shim calls https://github.com/rust-lang/rust/blob/a2cd91ceb0f156cb442d75e12dc77c3d064cdde4/library/std/src/rt.rs#L60 with the main function as first argument when using libstd. There are also a few other (unstable) options like marking a function with `#[start]`, in which case argc and argv are directly passed to that function without any additional arguments: https://github.com/bjorn3/rustc_codegen_cranelift/blob/3ea8915d4a247b5b3c4cfb3424c230ccd2645b17/example/alloc_example.rs#L30

Reply:
Should I open an issue for handling the main shim?

Reply:
> Should I open an issue for handling the main shim?

Yes @bjorn3 that would be great can you reference this PR in the issue too.

Reply:
#144

Reply:
Thanks for that @bjorn3